File: src/etch/examples/chat/Chatter.cs
using System;
using System.Collections.Generic;
using System.Text;

using Etch.Msg;
using Etch.Support;
using Etch.Transport;
using System.Globalization;
using Etch.Util;

namespace etch.examples.chat
{
    public class Chatter : ChatClient, Close
    {
        private ChatServer _server;
        private List<String> _whoIsOnline = new List<string>();
        private DateTimeFormatInfo _df = new DateTimeFormatInfo();

        public Chatter()
        {
            _df.MonthDayPattern = "dd/MM/yyyy";
        }

        public static void Main( String[] args )
        {
            String host;
            int port;

            if ( args.Length == 2 )
            {
                host = "127.0.0.1";
                port = int.Parse( args[ 1 ] );
            }
            else if ( args.Length == 0 )
            {
                host = "127.0.0.1";
                port = 4001;
            }
            else
            {
                Console.WriteLine( "usage: chatter host port" );
                Environment.Exit( 0 );
                throw new Exception();
            }

            //TcpConnection conn = new TcpConnection( null, host, port, 0 );
            //Chatter client = new Chatter();
            //Pool queuedPool = new QueuedPool();
            //Pool freePool = new FreePool();
            //ChatServer server = ChatHelper.NewChatServer( conn, client, queuedPool, freePool );
            //client.SetServer( server );

            //conn.Start();
            //conn.WaitUp( 4000 );

            // transport refactor
            URL uri = new URL( "tcp://127.0.0.1:4001" );
            Resources resources = ChatHelper.InitResources( null );

            Chatter client = new Chatter();
            RemoteChatServer server = ChatHelper.NewChatServer( uri, resources, client );
            client.SetServer( server );
            server._StartAndWaitUp( 4000 );

            Console.WriteLine( "( connected, enter commands or help)" );

            while ( true )
            {
                Console.Write( "cmd : " );
                Console.Out.Flush();

                String s = Console.ReadLine();
                if ( s.Length == 0 )
                    continue;

                char[] seps = { ' ', '\t', '\r', '\n' };
                String[] cmds = s.Split( seps );
                
                if ( cmds.Length == 0 )
                    continue;

                if ( cmds[ 0 ].Equals( "help" ) )
                {
                    Do_help();
                    continue;
                }

                if ( cmds[0].Equals( "login") )
                {
                    Do_login( server, cmds );
                    continue;
                }

                if ( cmds[0].Equals( "logout") )
                {
                    server.logout();
                    continue;
                }

                if ( cmds[0].Equals( "quit") )
                {
                    break;
                }

                if ( cmds[ 0 ].Equals( "send" ) )
                {
                    // extract message 
                    // note: cmds[2] = message to be extracted

                    if ( cmds.Length < 3 )
                    {
                        Console.WriteLine( "usage : login <username> <pwd>" );
                        continue;
                    }

                    char[] seps2 = {'\n', '\r'};
                    String msg = "";

                    for ( int i = 2; i < cmds.Length; i++ )
                        msg = msg + " " + cmds[ i ];

                    String[] cmds2 = msg.Split( seps2 );
                    Do_send( server, cmds[ 1 ], cmds2 );
                    continue;
                }

                if ( cmds[ 0 ].Equals( "who" ) )
                {
                    Do_who( client );
                    continue;
                }

                Console.WriteLine( "unknown command" );
                Do_help();
            }

            server.logout();
            ( ( RemoteChatServer ) server )._StopAndWaitDown( 4000 );
        }

        private static void Do_help()
        {
            Console.WriteLine( "commands: " );
	        Console.WriteLine( "  help" );
	        Console.WriteLine( "  login <user> <pwd>" );
	        Console.WriteLine( "  logout" );
	        Console.WriteLine( "  quit" );
 	        Console.WriteLine( "  send <user> <msg>" );
 	        Console.WriteLine( "  who" );
        }

        private static void Do_login( ChatServer server, String[] st )
        {
            try
            {
                String name = st[ 1 ];
                String pwd = st[ 2 ];
                server.login( name, pwd );
                Console.WriteLine( "login successful" );
            }
            catch ( etch.examples.chat.ConstsChat.Failure e )
            {
                Console.WriteLine( "failed : " + e.GetMessage() );
            }
            catch ( IndexOutOfRangeException )
            {
                Console.WriteLine( "usage : login <username> <pwd>" );
            }
            catch ( Exception e )
            {
                Console.WriteLine( e.Message );
            }
        }

        /// <summary>
        /// Send message 
        /// </summary>
        /// <param name="server"></param>
        /// <param name="user">username</param>
        /// <param name="st">message to be sent</param>
        /// 
        private static void Do_send( ChatServer server, String user, String[] st )
        {
            try
            {
                String who = user;
                String msg = st[ 0 ];
                server.send( who, msg );
            }
            catch ( etch.examples.chat.ConstsChat.Failure e )
            {
                Console.WriteLine( "failed : " + e.GetMessage() );
            }
            catch ( _Etch_AuthException )
            {
                Console.WriteLine( "you must login first" );
            }
            catch ( IndexOutOfRangeException )
            {
                Console.WriteLine( "usage: send <user> <msg> " );
            }
            catch ( Exception e )
            {
                
                Console.WriteLine( e.Message );
            }
        }

        private static void Do_who(Chatter client)
        {
            String[] names = client.GetWhoIsOnline();
            foreach ( String name in names )
            {
                Console.WriteLine( " " + name );
            }
        }

        private void SetServer( ChatServer server )
        {
            _server = server;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns>a list of who is currently online</returns>
        public String[] GetWhoIsOnline()
        {
            return _whoIsOnline.ToArray();
        }

        #region Close Members

        public void _Close( MessageSource src )
        {
            throw new Exception( "The method or operation is not implemented." );
        }

        #endregion

        #region ChatClient Members

        public void whoIsOnline( string[] names )
        {
            foreach ( String name in names )
            {
                Console.WriteLine( name + " is online" );
                _whoIsOnline.Add( name );
            }
        }

        public void statusChange( string name, bool? online )
        {
            if ( ( bool ) online )
            {
                Console.WriteLine( name + " is online" );
                _whoIsOnline.Add( name );
            }
            else
            {
                Console.WriteLine( name + " is offline" );
                _whoIsOnline.Remove( name );
            }
        }

        #endregion

        #region Chat Members

        public bool? isLoggedIn()
        {
            // server is always logged in to the client
            return true;
        }

        public void send( string who, string msg )
        {
            lock ( _df )
            {
                Console.WriteLine( "[ {0}@{1} ] {2}\n", who, DateTime.Today.ToString( _df ), msg );
            }
        }


        #endregion
    }
}

File: src/etch/examples/chat/Chatter.java
package etch.examples.chat;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.Date;
import java.util.HashSet;
import java.util.NoSuchElementException;
import java.util.Set;
import java.util.StringTokenizer;

import metreos.util.Log;
import metreos.util.URL;
import etch.bindings.java.support._Etch_AuthException;
import etch.bindings.java.transport.Session;

/**
 * Simple command line chat client.
 */
public class Chatter implements ChatClient, Session
{
	/**
	 * @param args
	 * @throws Exception 
	 */
	public static void main( String[] args ) throws Exception
	{
		Log.addSink( null );
		Log.report( "Chatter" );
		
		URL uri;
		if (args.length == 1)
		{
			uri = new URL( args[0] );
		}
		else if (args.length == 0)
		{
			uri = new URL( "tcp://localhost:4001" );
		}
		else
		{
			System.out.println( "usage: Chatter <uri>" );
			System.exit( 1 );
			return;
		}
		
		System.out.println( "Connecting to chat server at "+uri );

		Chatter client = new Chatter();
		RemoteChatServer server = ChatHelper.newChatServer( uri, null, client );
		server._startAndWaitUp( 4000 );
		
		System.out.println( "(connected, enter command or 'help' to list commands)" );
		
		BufferedReader rdr = new BufferedReader( new InputStreamReader( System.in ) );
		while (true)
		{
			System.out.print( "cmd: " );
			System.out.flush();
			
			String s = rdr.readLine();
			if (s.length() == 0)
				continue;
			
			StringTokenizer st = new StringTokenizer( s, " \t\r\n" );
			if (!st.hasMoreTokens())
				continue;
			
			String cmd = st.nextToken();
			
			if (cmd.equals( "help" ))
			{
				do_help();
				continue;
			}
			
			if (cmd.equals( "login" ))
			{
				do_login( server, st );
				continue;
			}
			
			if (cmd.equals( "logout" ))
			{
				do_logout( server );
				continue;
			}
			
			if (cmd.equals( "quit" ))
				break;
			
			if (cmd.equals( "send" ))
			{
				do_send( server, cmd, st );
				continue;
			}
			
			if (cmd.equals( "who" ))
			{
				do_who( client );
				continue;
			}

			System.out.println( "unknown command "+cmd );
			do_help();
		}
		
		server._stop();
	}

	private static void do_help()
    {
	    System.out.println( "commands: " );
	    System.out.println( "  help" );
	    System.out.println( "  login <user> <pwd>" );
	    System.out.println( "  logout" );
	    System.out.println( "  quit" );
 	    System.out.println( "  send <user> <msg>" );
 	    System.out.println( "  who" );
	}

	private static void do_login( ChatServer server, StringTokenizer st )
    {
		try
		{
			String name = st.nextToken();
			String pwd = st.nextToken();
			server.login( name, pwd );
			System.out.println( "(login successful)" );
		}
		catch ( NoSuchElementException e )
		{
			System.out.println( "usage: login <user> <pwd>" );
		}
		catch ( Failure e )
		{
			System.out.println( "failed: "+e.getMessage() );
			e.printStackTrace();
		}
		catch ( Exception e )
		{
			System.out.println( "failed: "+e );
			e.printStackTrace();
		}
    }

	private static void do_logout( ChatServer server )
	{
		server.logout();
		System.out.println( "(logout successful)" );
	}

	private static void do_send( ChatServer server, String cmd, StringTokenizer st )
    {
		try
		{
			String who = st.nextToken();
			String msg = st.nextToken( "\r\n" );
			server.send( who, msg );
		}
		catch ( NoSuchElementException e )
		{
			System.out.println( "usage: send <user> <msg>" );
		}
		catch ( _Etch_AuthException e )
		{
			System.out.println( "you must login first" );
		}
		catch ( Failure e )
		{
			System.out.println( "failed: "+e.getMessage() );
		}
		catch ( Exception e )
		{
			System.out.println( "failed: "+e );
		}
    }
	
	private static void do_who( Chatter client )
    {
	    for (String name: client.getWhoIsOnline())
	    	System.out.println( "  "+name );
    }
	
	/**
	 * @return a list of who is online right now.
	 */
	public String[] getWhoIsOnline()
	{
		return whoIsOnline.toArray( new String[0] );
	}
	
	private final Set<String> whoIsOnline = Collections.synchronizedSet( new HashSet<String>() );

    ////////////////
    // ChatClient //
    ////////////////

	public void whoIsOnline( String[] names )
    {
	    for (String name: names)
	    {
	    	whoIsOnline.add( name );
	    	System.out.println( name+" is online" );
	    }
    }

	public void statusChange( String name, Boolean online )
    {
		if (online)
		{
			whoIsOnline.add( name );
			System.out.println( name+" is now online" );
		}
		else
		{
			whoIsOnline.remove( name );
			System.out.println( name+" is now offline" );
		}
    }

	//////////
	// Chat //
	//////////

	public Boolean isLoggedIn()
    {
		// server is always logged in to the client.
	    return true;
    }

	public void send( String who, String msg ) throws Failure
    {
		synchronized( df )
		{
			System.out.printf( "[ %s@%s ] %s\n", who, df.format( new Date() ), msg );
		}
    }
	
	private DateFormat df = new SimpleDateFormat( "HH:mm:ss" );

	////////////////
	// OOBHandler //
	////////////////

	public Object sessionQuery( Object query )
	{
		throw new UnsupportedOperationException( "unknown query: "+query );
	}

	public void sessionControl( Object control, Object value )
	{
		throw new UnsupportedOperationException( "unknown control: "+control );
	}

	public void sessionNotify( Object event )
	{
		if (event.equals( Session.UP ) )
		{
			System.out.println( "(server opened connection)" );
			return;
		}
		
		if (event.equals( Session.DOWN ) )
		{
			System.out.println( "(server closed connection)" );
			return;
		}
		
		if (event instanceof Throwable)
		{
			((Throwable) event).printStackTrace();
			return;
		}
	}
}
